### PR TITLE
8306825: Monitor deflation might be accidentally disabled by zero intervals

### PR DESCRIPTION
**Pull request by shipilev**
Found this mistake while backporting [JDK-8305994](https://bugs.openjdk.org/browse/JDK-8305994): if we put GuaranteedAsyncDeflationInterval=0 (defined as "disable guaranteed deflation"), then the MonitorDeflationThread would call `wait(0)`, that is, wait indefinitely for the notify. This breaks triggering the threshold heuristics!

But what is even more concerning, the same thing would happen even prior to [JDK-8305994](https://bugs.openjdk.org/browse/JDK-8305994), if we put GuaranteedSafepointInterval=0 too. Which means users who kept GSI=0 as the way to dodge excess safepoints, now effectively disable monitor deflation!

New test cases capture most of the interesting paths/combinations. Since I needed to add more test cases, I ended up refactoring the test for more clarity. The test is also renamed, because now it tests not only GADI, but other intervals too.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8306825](https://bugs.openjdk.org/browse/JDK-8306825): Monitor deflation might be accidentally disabled by zero intervals

### Reviewers
 * [Daniel D. Daugherty](https://openjdk.org/census#dcubed) (@dcubed-ojdk - **Reviewer**) ⚠️ Review applies to [4a60694f](https://git.openjdk.org/jdk/pull/13635/files/4a60694fb973cbfe0e1a1211c69eb9d8d5d3d6f9)
 * [Evgeny Astigeevich](https://openjdk.org/census#eastigeevich) (@eastig - Committer) ⚠️ Review applies to [4a60694f](https://git.openjdk.org/jdk/pull/13635/files/4a60694fb973cbfe0e1a1211c69eb9d8d5d3d6f9)
 * [Paul Hohensee](https://openjdk.org/census#phh) (@phohensee - **Reviewer**) ⚠️ Review applies to [4a60694f](https://git.openjdk.org/jdk/pull/13635/files/4a60694fb973cbfe0e1a1211c69eb9d8d5d3d6f9)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/13635/head:pull/13635` \
`$ git checkout pull/13635`

Update a local copy of the PR: \
`$ git checkout pull/13635` \
`$ git pull https://git.openjdk.org/jdk.git pull/13635/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 13635`

View PR using the GUI difftool: \
`$ git pr show -t 13635`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/13635.diff">https://git.openjdk.org/jdk/pull/13635.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/13635#issuecomment-1523856929)

### CODE REVIEW

**Comment by shipilev** [comment]
@dcubed-ojdk, you might be interested in the continuation of this small saga :)

File: test/hotspot/jtreg/runtime/Monitor/DeflationIntervalsTest.java [hunk 1]
```diff
@@ -0,0 +1,357 @@
...
+                // Try with all defaults
+                test(Disabled.NO, Guaranteed.MAYBE, Threshold.MAYBE);
+                break;
+
+            case "allIntervalsZero":
+                // Try with all deflation intervals at zero
+                test(Disabled.YES, Guaranteed.NO, Threshold.NO,
+                    "-XX:GuaranteedAsyncDeflationInterval=0",
+                    "-XX:AsyncDeflationInterval=0",
+                    "-XX:MonitorUsedDeflationThreshold=0",
```
Comment (dcubed-ojdk) [inline test/hotspot/jtreg/runtime/Monitor/DeflationIntervalsTest.java:174]: But... this one isn't an interval option...

File: test/hotspot/jtreg/runtime/Monitor/DeflationIntervalsTest.java [hunk 2]
```diff
@@ -0,0 +1,357 @@
...
+                    "-XX:GuaranteedSafepointInterval=0"
+                );
+                break;
+
+            case "allThresholdsZero":
+                // Try with all heuristics thresholds at zero
+                test(Disabled.NO, Guaranteed.NO, Threshold.NO,
+                    "-XX:GuaranteedAsyncDeflationInterval=0",
```
Comment (dcubed-ojdk) [inline test/hotspot/jtreg/runtime/Monitor/DeflationIntervalsTest.java:182]: But... this one isn't a threshold option...

**Review by dcubed-ojdk — APPROVED** [review]
Just a couple of questions about the test. Thumbs up.
I'll do another round of Mach5 testing for this fix.

File: test/hotspot/jtreg/runtime/Monitor/DeflationIntervalsTest.java (hunk 1 above)
Comment (shipilev) [reply test/hotspot/jtreg/runtime/Monitor/DeflationIntervalsTest.java:174]: Right. Dropped! Has no effect on the test outcome.

File: test/hotspot/jtreg/runtime/Monitor/DeflationIntervalsTest.java (hunk 2 above)
Comment (shipilev) [reply test/hotspot/jtreg/runtime/Monitor/DeflationIntervalsTest.java:182]: Right, dropped. Now we have `Guaranteed.MAYBE` on that path :)

**Review by dcubed-ojdk — APPROVED** [review]

**Comment by shipilev** [comment]
Do you want to test a new version of this fix, @dcubed-ojdk? The one that dropped two options from two configs? I am pretty sure those changes are innocuous, but...

**Comment by dcubed-ojdk** [comment]
Yup. I just kicked off a Mach5 Tier1 and I'll do the Repeat-50 runs...

**Comment by dcubed-ojdk** [comment]
Mach5 Tier1 had no failures.

Mach5 Repeat-50 of runtime/Monitor/DeflationIntervalsTest.java also had no failures:
- linux-aarch64, linux-x64, macosx-aarch64, macosx-x64, windows-x64
- 50 iterations of all 11 sub-tests passed on all 5 platforms

**Comment by shipilev** [comment]
Excellent, thank you, @dcubed-ojdk!

Anyone else for (R)eview?

File: src/hotspot/share/runtime/monitorDeflationThread.cpp [hunk 3]
```diff
@@ -48,9 +48,36 @@ void MonitorDeflationThread::initialize() {
...
+  // We wait for the lowest of these three intervals:
+  //  - GuaranteedSafepointInterval
+  //      While deflation is not related to safepoint anymore, this keeps compatibility with
+  //      the old behavior when deflation also happened at safepoints. Users who set this
+  //      option to get more/less frequent deflations would be served with this option.
+  //  - AsyncDeflationInterval
+  //      Normal threshold-based deflation heuristics checks the conditions at this interval.
+  //      See is_async_deflation_needed().
+  //  - GuaranteedAsyncDeflationInterval
+  //      Backup deflation heuristics checks the conditions at this interval.
```
Comment (eastig) [inline src/hotspot/share/runtime/monitorDeflationThread.cpp:60]: Either `heuristics check` or `heuristic checks`.

**Review by eastig — APPROVED** [review]
lgtm, a minor grammar comment.

**Review by phohensee — APPROVED** [review]
Lgtm.

File: src/hotspot/share/runtime/monitorDeflationThread.cpp (hunk 3 above)
Comment (shipilev) [reply src/hotspot/share/runtime/monitorDeflationThread.cpp]: Fixed, thanks.

**Comment by shipilev** [comment]
Thank you all!

/integrate

**Pull request closed without merging**